blosc/shuffle.c

```diff
@@ -13,6 +13,37 @@
 #include "blosc-comp-features.h"
 #include <stdio.h>
 
+#if defined(_WIN32)
+
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+
+#include <windows.h>
+
+/**
+ * pthread_once implementation based on the MS Windows One-Time Initialization
+ * (https://docs.microsoft.com/en-us/windows/desktop/Sync/one-time-initialization)
+ * APIs.
+ */
+typedef INIT_ONCE pthread_once_t;
+#define PTHREAD_ONCE_INIT INIT_ONCE_STATIC_INIT
+static int pthread_once(pthread_once_t* once_control,
+                        void (*init_routine)(void)) {
+  BOOL pending;
+  InitOnceBeginInitialize(once_control, /*dwFlags=*/0, /*fPending=*/&pending,
+                          NULL);
+  if (pending == TRUE) {
+    init_routine();
+    InitOnceComplete(once_control, /*dwFlags=*/0, /*lpContext=*/NULL);
+  }
+  return 0;
+}
+
+#else
+#include <pthread.h>
+#endif
+
```

Comment:
For consistency, I think it be worth adding `pthread_once()` to https://github.com/Blosc/c-blosc/blob/master/blosc/win32/pthread.c instead.  Would you agree?

Reply:
Also, do you think this would be compatible with at least VS2008?  We need that for Python2 libraries that embed C-Blosc.

Reply:
Adding it to pthread.h/pthread.c instead of inlining it into shuffle.c makes sense.  However, I am a bit confused by the fact that blosc.c #includes pthread.c, but in addition the CMakeLists.txt file sometimes includes win32/pthread.c in the SOURCES (which would lead to duplicate definitions).  I guess that condition (the system Threads library not being found) must never be satisfied in practice?  Wouldn't it make more sense to remove the #include pthread.c from blosc.c, and just rely on the build configuration to include win32/pthread.c ?

Regarding VS2008, I'm not sure what the interaction is between VS2008 and Windows API versions.  I don't have a way to test the Windows build other than Appveyor, and I see that the appveyor configuration for this package doesn't seem to test with VS2008.

Reply:
Another issue that occurred to me is that there could be a conflict if on Windows the user links c-blosc, with its own partial pthread library that exports some pthread_* symbols, and another more full-featured pthread library that also exports some pthread_* symbols.  I guess this would only be an issue if static linking is used on Windows.  The solution to that problem would be to add to win32/pthread.h #define pthread_XXX blosc_internal_pthread_XXX  for each pthread symbol.

Reply:
> Adding it to pthread.h/pthread.c instead of inlining it into shuffle.c makes sense. However, I am a bit confused by the fact that blosc.c #includes pthread.c, but in addition the CMakeLists.txt file sometimes includes win32/pthread.c in the SOURCES (which would lead to duplicate definitions). I guess that condition (the system Threads library not being found) must never be satisfied in practice? Wouldn't it make more sense to remove the #include pthread.c from blosc.c, and just rely on the build configuration to include win32/pthread.c ?

Yes, AFAIK there is not official port of pthreads for Windows, so it is safe to assume that with win32/VS will never be double inclusion of the pthread.c and win32/pthread.c.  This has been working for years, and frankly, I would not touch this configuration.

> 
> Regarding VS2008, I'm not sure what the interaction is between VS2008 and Windows API versions. I don't have a way to test the Windows build other than Appveyor, and I see that the appveyor configuration for this package doesn't seem to test with VS2008.

That's correct, there is no provision for testing VS2008 in Appveyor.  This should be fixed somehow (honestly, I don't remember the reason why VS2008 is not tested anymore).

Reply:
> Another issue that occurred to me is that there could be a conflict if on Windows the user links c-blosc, with its own partial pthread library that exports some pthread_* symbols, and another more full-featured pthread library that also exports some pthread_* symbols. I guess this would only be an issue if static linking is used on Windows. The solution to that problem would be to add to win32/pthread.h #define pthread_XXX blosc_internal_pthread_XXX for each pthread symbol.

As I said before, there is not a 'blessed' version of pthreads for Windows (mainly because Windows comes with its own threading library), so I don't think we should worry about potential duplicated symbols on the static Windows version of the library.